Implement DualRunning state to enable blue-green deploys

glaksh100 opened the pull request:
This PR implements the state machine required for a blue green deployment. It's huge but almost all of the changes to existing states are gated on the `isBlueGreenDeployment` check. Pardon the long commit log, rebasing (to squash) was too hellish.

For a review, looking at the following will help:

1. Additions to FlinkApplicationStatus:
   - UpdatingVersion: blue/green
   - DeployingVersion: blue/green
   - UpdatingHash
   - TeardownHash

1. flink_state_machine changes:
   - NewOrUpdating: If BG, updates the UpdatingVersion/Hash etc.
   - ClusterStarting: If BG, updates the version and hash on the `FlinkApplicationVersionStatus`. If  BG and `SavepointDisabled`, transitions directly to `SubmittingJob`
   - Savepointing: If BG, does not cancel job and also does not reset Job ID to "" (as the old job is still running)
   - SubmittingJob: If BG (and not first deploy), transitions to `DualRunning` 
   - DualRunning: State in which two jobs are running, cluster/job status get updated and application transitions out of this state when a `Teardown` flag is set. 
   - Teardown: State in which the current `DeployVersion` is torn down and the `UpdatingVersion` becomes the `DeployVersion`
   - Deleting: If BG, assume that there could be two jobs running and cance/delete both jobs. 

1. Flink controller changes:
   - A couple of API changes: `CancelWithSavepoint` --> `Savepoint` with cancelFlag. `Savepoint` and `GetSavepointStatus` accept a JobID (mainly for purposes of deletion)
   - Helper methods for Deleting resources during teardown

1. Other noteworthy changes:
   - Generic service now has a version suffix
   - During a teardown, the teardownHash is noted and `HashForApplication` now includes FailedHash/TeardownHash in its exemption (so as to not trigger continuous updates during a teardown)
  
**Testing**
- Unit tests for the controllers/state machine
- Integration test for the entire Running --> DualRunning --> Teardown cycle
- Verifying locally that Dual deployments are not impacted by this change (both status sub-resource and otherwise)
- Verifying BlueGreen deployment mode locally

**TODO**
- Documentaion updates: I'll do that once we get through the code review.

glaksh100 commented:
@mwylde @anandswaminathan This is ready for an initial review. I've added some pointers for review.  Hopefully the last long PR for this feature, and as always, TIA for your feedback :)

glaksh100 commented on pkg/controller/flinkapplication/flink_state_machine.go:
```diff
@@ -867,6 +916,192 @@ func (s *FlinkStateMachine) compareAndUpdateError(application *v1beta1.FlinkAppl
+	return statusUnchanged, nil
+}
+
+func (s *FlinkStateMachine) handleTeardown(ctx context.Context, application *v1beta1.FlinkApplication) (bool, error) {
```
There are a couple of things with the teardown phase that I'd like opinions on:
1) The Teardown spec. Right now, I've modeled it exactly like the `forceRollback` flag (which is a boolean). However, I've been thinking that this may have to actually be the version we need to teardown, simply because in the event of a bad update, the user may actually want to teardown the new(er) version. 

2) The act of teardown: Right now the cluster is torn down immediately. Should this instead try to savepoint/force cancel the job? I'm not too sure what the benefit of savepointing during a teardown would be, but force-cancelling perhaps would help in a more graceful deletion of the job/cluster.

Let me know what you think!

mwylde replied:
1. Commented a bit about this with a slightly different concern (that people will forget to unset it, leading to unintentional tear-downs). But the one you bring up is also important: if the green deploy is bad, we need to give users a way to tear down that one.
2. I think this definitely should be a ForceCancel, so that we don't cause ZK issues.

glaksh100 commented on pkg/controller/flink/container_utils.go:
```diff
@@ -34,7 +34,7 @@ const (
 	RestartNonce                     = "restart-nonce"
 	FlinkApplicationVersionEnv       = "FLINK_APPLICATION_VERSION"
-	FlinkApplicationVersion          = "flink-application-version"
+	FlinkApplicationVersion          = "flink-app-version"
```
You're right, I was trying to make it consistent with `flink-app-hash` but this is not worth the redeployment.

glaksh100 commented on pkg/controller/flink/client/api.go:
```diff
@@ -47,7 +47,7 @@ const programInvocationException = "org.apache.flink.client.program.ProgramInvoc
 
 type FlinkAPIInterface interface {
-	CancelJobWithSavepoint(ctx context.Context, url string, jobID string) (string, error)
+	CancelJobWithSavepoint(ctx context.Context, url string, jobID string, cancel bool) (string, error)
```
I separated this into two methods: `CancelJobWithSavepoint` and `SavepointJob`. Thanks for the suggestion!

glaksh100 commented:
@mwylde I've addressed all of the review comments as per the discussion we had offline. The changes include:
- Merging everything into the `DualRunning` state
- Using a `tearDownVersionHash` in the spec, which would specify which version needs to be torn down (could be either of the two)
- Adding a suffix for events like e.g. `Creating Flink cluster for deploy 3c671670 for version blue` (I tried to make this least invasive in terms of code)
- Cleaning up resources when we enter a DeployFailed with BG mode
- Preventing deploymentMode switches by storing the info in a `DeploymentMode` within the `Status`

Let me know what you think and I'll be sure to add docs along with any final review comments.

glaksh100 commented:
I've now updated docs and the state machine diagrams too!
